File: api-node/src/routes/checkins.ts
import { Router, Response } from "express";
import { getCheckInsCollection, getBreweriesCollection, getUsersCollection } from "../db";
import { randomUUID } from "crypto";
import { authMiddleware, AuthenticatedRequest } from "../middleware/auth";

const router = Router();

// Haversine Distance helper (coords: [longitude, latitude])
function haversineDistance(coords1: [number, number], coords2: [number, number]): number {
  const toRad = (x: number) => (x * Math.PI) / 180;
  const [lon1, lat1] = coords1;
  const [lon2, lat2] = coords2;

  const R = 3958.8; // Earth radius in miles
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

// Sequential distance calculator within a trip (uses home_coordinates for first stop if present)
export async function recalculateTripDistances(userId: string, tripName: string | null): Promise<void> {
  if (!tripName || !tripName.trim()) return;

  const checkinsCol = getCheckInsCollection();
  const breweriesCol = getBreweriesCollection();
  const usersCol = getUsersCollection();

  // Fetch and sort chronologically (visited_at ascending)
  const checkins = await checkinsCol
    .find({ user_id: userId, trip_name: tripName.trim() })
    .sort({ visited_at: 1 })
    .toArray();

  if (checkins.length === 0) return;

  // Resolve coordinate map
  const breweryIds = checkins.map((c) => c.brewery_id);
  const breweries = await breweriesCol.find({ id: { $in: breweryIds } }).toArray();
  const breweriesMap = new Map<string, [number, number]>();
  breweries.forEach((b) => {
    if (b.location && b.location.coordinates) {
      breweriesMap.set(b.id, b.location.coordinates);
    }
  });

  // Resolve user home coordinates for the first stop fallback
  const user = await usersCol.findOne({ id: userId });
  const homeCoords = user?.home_coordinates;

  // Calculate distances
  for (let i = 0; i < checkins.length; i++) {
    const current = checkins[i];
    let distance = 0;

    if (i > 0) {
      const prev = checkins[i - 1];
      const coords1 = breweriesMap.get(prev.brewery_id);
      const coords2 = breweriesMap.get(current.brewery_id);

      if (coords1 && coords2) {
        distance = haversineDistance(coords1, coords2);
      }
    } else if (homeCoords) {
      // First stop: calculate distance from user's home location coordinates!
      const currentCoords = breweriesMap.get(current.brewery_id);
      if (currentCoords) {
        distance = haversineDistance(homeCoords as [number, number], currentCoords);
      }
    }

    const roundedDistance = parseFloat(distance.toFixed(2));
    await checkinsCol.updateOne(
      { id: current.id },
      { $set: { distance_miles: roundedDistance } }
    );
  }
}

// GET /checkins - Get user check-in history sorted by visited_at descending (Authenticated and scoped to user)
router.get("/", authMiddleware as any, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const user_id = req.user!.id;
    const trip_name = req.query.trip_name as string | undefined;
    const limitParam = req.query.limit as string | undefined;

    let limit = 50;
    if (limitParam) {
      const parsed = parseInt(limitParam, 10);
      if (isNaN(parsed) || parsed < 1) {
        return res.status(400).json({
          code: "BAD_REQUEST",
          message: "The 'limit' parameter must be a positive integer."
        });
      }
      limit = parsed;
    }

    const filter: any = { user_id };
    if (trip_name) {
      filter.trip_name = { $regex: trip_name, $options: "i" };
    }

    const checkins = await getCheckInsCollection()
      .find(filter, { projection: { _id: 0 } })
      .sort({ visited_at: -1 })
      .limit(limit)
      .toArray();

    return res.json(checkins);
  } catch (error: any) {
    console.error("Error fetching checkins:", error);
    return res.status(500).json({
      code: "INTERNAL_SERVER_ERROR",
      message: error.message || "An unexpected error occurred."
    });
  }
});

// POST /checkins - Log a new check-in (Authenticated and set to user_id)
router.post("/", authMiddleware as any, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const {
      brewery_id,
      visited_at,
      rating,
      took_tour,
      notes,
      distance_miles,
      transportation_mode,
      trip_name,
      amenities_observed
    } = req.body;

    const user_id = req.user!.id;

    // Body Validation
    if (!brewery_id || typeof brewery_id !== "string") {
      return res.status(400).json({ code: "BAD_REQUEST", message: "brewery_id is required." });
    }
    if (!visited_at || isNaN(Date.parse(visited_at))) {
      return res.status(400).json({ code: "BAD_REQUEST", message: "visited_at must be a valid ISO datetime string." });
    }
    if (rating === undefined || typeof rating !== "number" || rating < 1 || rating > 5) {
      return res.status(400).json({ code: "BAD_REQUEST", message: "rating must be a number between 1 and 5." });
    }
    if (took_tour === undefined || typeof took_tour !== "boolean") {
      return res.status(400).json({ code: "BAD_REQUEST", message: "took_tour must be a boolean." });
    }
    if (notes === undefined || typeof notes !== "string") {
      return res.status(400).json({ code: "BAD_REQUEST", message: "notes must be a string." });
    }
    if (distance_miles === undefined || typeof distance_miles !== "number" || distance_miles < 0) {
      return res.status(400).json({ code: "BAD_REQUEST", message: "distance_miles must be a positive number." });
    }
    
    const validModes = ["drive", "flight", "walk", "transit"];
    if (!transportation_mode || !validModes.includes(transportation_mode)) {
      return res.status(400).json({
        code: "BAD_REQUEST",
        message: `transportation_mode must be one of: ${validModes.join(", ")}`
      });
    }

    if (trip_name !== undefined && trip_name !== null && typeof trip_name !== "string") {
      return res.status(400).json({ code: "BAD_REQUEST", message: "trip_name must be a string or null." });
    }

    if (!Array.isArray(amenities_observed) || !amenities_observed.every(item => typeof item === "string")) {
      return res.status(400).json({ code: "BAD_REQUEST", message: "amenities_observed must be an array of strings." });
    }

    // Verify the brewery exists
    const brewery = await getBreweriesCollection().findOne({ id: brewery_id });
    if (!brewery) {
      return res.status(404).json({
        code: "NOT_FOUND",
        message: `Brewery with ID ${brewery_id} not found.`
      });
    }

    const cleanTripName = trip_name ? trip_name.trim() : null;

    let initialDistance = 0;
    if (!cleanTripName) {
      const user = await getUsersCollection().findOne({ id: user_id });
      if (user && user.home_coordinates && brewery && brewery.location && brewery.location.coordinates) {
        initialDistance = haversineDistance(user.home_coordinates as [number, number], brewery.location.coordinates as [number, number]);
      }
    }

    const newCheckIn = {
      id: randomUUID(),
      user_id,
      brewery_id,
      visited_at: new Date(visited_at).toISOString(),
      rating,
      took_tour,
      notes,
      distance_miles: parseFloat(initialDistance.toFixed(2)),
      transportation_mode,
      trip_name: cleanTripName,
      amenities_observed
    };

    const checkinsCol = getCheckInsCollection();
    await checkinsCol.insertOne(newCheckIn);

    // Recalculate trip distances if part of a trip
    if (cleanTripName) {
      await recalculateTripDistances(user_id, cleanTripName);
    }

    const { _id, ...responseBody } = newCheckIn as any;
    return res.status(201).json(responseBody);
  } catch (error: any) {
    console.error("Error creating checkin:", error);
    return res.status(500).json({
      code: "INTERNAL_SERVER_ERROR",
      message: error.message || "An unexpected error occurred."
    });
  }
});

// PUT /checkins/:id - Update trip_name or visited_at on an existing check-in (Authenticated & Scoped)
router.put("/:id", authMiddleware as any, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const id = req.params.id;
    const user_id = req.user!.id;
    const { trip_name, visited_at } = req.body;

    const checkinsCol = getCheckInsCollection();
    const existing = await checkinsCol.findOne({ id, user_id });

    if (!existing) {
      return res.status(404).json({
        code: "NOT_FOUND",
        message: `Check-in record with ID ${id} not found for this user.`
      });
    }

    const oldTripName = existing.trip_name;
    const newTripName = trip_name !== undefined ? (trip_name ? trip_name.trim() : null) : oldTripName;
    const newVisitedAt = visited_at ? new Date(visited_at).toISOString() : existing.visited_at;

    await checkinsCol.updateOne(
      { id, user_id },
      { $set: { trip_name: newTripName, visited_at: newVisitedAt } }
    );

    // Recalculate routing distance layers
    if (oldTripName) {
      await recalculateTripDistances(user_id, oldTripName);
    }
    if (newTripName && newTripName !== oldTripName) {
      await recalculateTripDistances(user_id, newTripName);
    }

    const updated = await checkinsCol.findOne({ id, user_id }, { projection: { _id: 0 } });
    return res.json(updated);
  } catch (error: any) {
    console.error("Error updating check-in:", error);
    return res.status(500).json({
      code: "INTERNAL_SERVER_ERROR",
      message: error.message || "An unexpected error occurred."
    });
  }
});

// PATCH /checkins/assign-trip - Batch update to group/ungroup check-ins in a single request (Authenticated)
router.patch("/assign-trip", authMiddleware as any, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const user_id = req.user!.id;
    const { checkin_ids, trip_name } = req.body;

    if (!Array.isArray(checkin_ids) || checkin_ids.some(item => typeof item !== "string")) {
      return res.status(400).json({
        code: "BAD_REQUEST",
        message: "checkin_ids must be an array of string UUIDs."
      });
    }

    const cleanTripName = trip_name ? trip_name.trim() : null;
    const checkinsCol = getCheckInsCollection();

    // Fetch affected check-ins to capture their pre-existing trip names
    const affected = await checkinsCol
      .find({ id: { $in: checkin_ids }, user_id })
      .toArray();

    const oldTrips = Array.from(new Set(affected.map((c) => c.trip_name).filter(Boolean)));

    // Batch update
    await checkinsCol.updateMany(
      { id: { $in: checkin_ids }, user_id },
      { $set: { trip_name: cleanTripName } }
    );

    // Trigger sequential Haversine updates
    if (cleanTripName) {
      await recalculateTripDistances(user_id, cleanTripName);
    }
    for (const oldTrip of oldTrips) {
      if (oldTrip !== cleanTripName) {
        await recalculateTripDistances(user_id, oldTrip);
      }
    }

    return res.json({
      message: "Check-ins retroactively assigned and Haversine distances synchronized successfully."
    });
  } catch (error: any) {
    console.error("Error in batch check-in assignment:", error);
    return res.status(500).json({
      code: "INTERNAL_SERVER_ERROR",
      message: error.message || "An unexpected error occurred."
    });
  }
});

export default router;

File: api-python/routers/checkins.py
import uuid
import math
from typing import List, Optional
from pydantic import BaseModel
from fastapi import APIRouter, HTTPException, Query, status, Depends
from bson import ObjectId
from database import get_db
from models import CheckIn, CheckInCreate
from routers.auth import get_current_user

router = APIRouter(prefix="/checkins", tags=["Check-Ins"])

class AssignTripPayload(BaseModel):
    checkin_ids: List[str]
    trip_name: Optional[str] = None

class UpdateCheckInPayload(BaseModel):
    trip_name: Optional[str] = None
    visited_at: Optional[str] = None

# Haversine Distance helper (coords: [longitude, latitude])
def haversine_distance(coords1: List[float], coords2: List[float]) -> float:
    lon1, lat1 = coords1
    lon2, lat2 = coords2
    
    # Convert decimal degrees to radians
    lon1, lat1, lon2, lat2 = map(math.radians, [lon1, lat1, lon2, lat2])
    
    # Haversine formula
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = math.sin(dlat/2)**2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon/2)**2
    c = 2 * math.asin(math.sqrt(a))
    r = 3958.8  # Radius of earth in miles
    return c * r

# Sequential distance calculator within a trip (uses home_coordinates for first stop if present)
async def recalculate_trip_distances(user_id: str, trip_name: Optional[str]):
    if not trip_name or not trip_name.strip():
        return

    db = get_db()
    trip_name_clean = trip_name.strip()

    # Fetch and sort chronologically (visited_at ascending)
    cursor = db.checkins.find({"user_id": user_id, "trip_name": trip_name_clean}).sort("visited_at", 1)
    checkins = await cursor.to_list(length=None)

    if not checkins:
        return

    # Resolve coordinates
    brewery_ids = [c["brewery_id"] for c in checkins]
    brew_cursor = db.breweries.find({"id": {"$in": brewery_ids}})
    breweries = await brew_cursor.to_list(length=None)
    breweries_map = {b["id"]: b["location"]["coordinates"] for b in breweries if b.get("location")}

    # Resolve user home coordinates for the first stop fallback with resilient user lookup
    user_query = {"$or": [{"id": user_id}, {"_id": user_id}]}
    try:
        if ObjectId.is_valid(user_id):
            user_query["$or"].append({"_id": ObjectId(user_id)})
    except Exception:
        pass
    user = await db.users.find_one(user_query)
    if not user:
        print(f"[WARN] Could not find user document in database for user_id: {user_id}")
    home_coords = user.get("home_coordinates") if user else None

    # Calculate and update
    for i, current in enumerate(checkins):
        distance = 0.0
        if i > 0:
            prev = checkins[i - 1]
            coords1 = breweries_map.get(prev["brewery_id"])
            coords2 = breweries_map.get(current["brewery_id"])
            if coords1 and coords2:
                distance = haversine_distance(coords1, coords2)
        elif home_coords:
            # First stop: calculate distance from user's home location coordinates!
            current_coords = breweries_map.get(current["brewery_id"])
            if current_coords:
                distance = haversine_distance(home_coords, current_coords)

        rounded_distance = round(distance, 2)
        await db.checkins.update_one(
            {"id": current["id"]},
            {"$set": {"distance_miles": rounded_distance}}
        )

@router.get("", response_model=List[CheckIn])
async def get_checkins(
    trip_name: Optional[str] = None,
    limit: int = Query(50, ge=1),
    current_user: dict = Depends(get_current_user)
):
    db = get_db()
    user_id = current_user["id"]
    filter_query = {"user_id": user_id}

    if trip_name:
        filter_query["trip_name"] = {"$regex": trip_name, "$options": "i"}

    cursor = db.checkins.find(filter_query, {"_id": 0}).sort("visited_at", -1).limit(limit)
    checkins = await cursor.to_list(length=limit)
    return checkins

@router.post("", response_model=CheckIn, status_code=status.HTTP_201_CREATED)
async def create_checkin(
    payload: CheckInCreate,
    current_user: dict = Depends(get_current_user)
):
    db = get_db()
    user_id = current_user["id"]

    # Verify the brewery exists
    brewery = await db.breweries.find_one({"id": payload.brewery_id})
    if not brewery:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Brewery with ID {payload.brewery_id} not found."
        )

    # Format and validate visited_at to ISO string
    try:
        from datetime import datetime
        iso_str = payload.visited_at
        if iso_str.endswith("Z"):
            iso_str = iso_str[:-1] + "+00:00"
        datetime.fromisoformat(iso_str)
        formatted_visited_at = payload.visited_at
        if not formatted_visited_at.endswith("Z") and "+" not in formatted_visited_at:
            formatted_visited_at += "Z"
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="visited_at must be a valid ISO datetime string."
        )

    clean_trip_name = payload.trip_name.strip() if payload.trip_name else None

    # Resolve home coordinates and calculate distance for single/standalone stops
    initial_distance = 0.0
    if not clean_trip_name:
        user_query = {"$or": [{"id": user_id}, {"_id": user_id}]}
        try:
            if ObjectId.is_valid(user_id):
                user_query["$or"].append({"_id": ObjectId(user_id)})
        except Exception:
            pass
        user = await db.users.find_one(user_query)
        if user:
            home_coords = user.get("home_coordinates")
            if home_coords and isinstance(home_coords, list) and len(home_coords) == 2 and brewery.get("location") and brewery["location"].get("coordinates"):
                initial_distance = haversine_distance(home_coords, brewery["location"]["coordinates"])
            else:
                print(f"[WARN] User {user.get('email')} missing home_coordinates")
        else:
            print(f"[WARN] Could not find user document in database for user_id: {user_id}")

    new_checkin = payload.dict()
    new_checkin["id"] = str(uuid.uuid4())
    new_checkin["user_id"] = user_id  # set authenticated user
    new_checkin["visited_at"] = formatted_visited_at
    new_checkin["trip_name"] = clean_trip_name
    new_checkin["distance_miles"] = round(initial_distance, 2)

    await db.checkins.insert_one(new_checkin)

    # Recalculate trip distances if part of a trip
    if clean_trip_name:
        await recalculate_trip_distances(user_id, clean_trip_name)

    # Query and return the freshly updated check-in document in the HTTP response
    fresh_checkin = await db.checkins.find_one({"id": new_checkin["id"]}, {"_id": 0})
    if not fresh_checkin:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve the newly registered check-in."
        )

    return fresh_checkin

@router.put("/{id}", response_model=CheckIn)
async def update_checkin(
    id: str,
    payload: UpdateCheckInPayload,
    current_user: dict = Depends(get_current_user)
):
    db = get_db()
    user_id = current_user["id"]

    existing = await db.checkins.find_one({"id": id, "user_id": user_id})
    if not existing:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Check-in record with ID {id} not found."
        )

    old_trip_name = existing.get("trip_name")
    
    # Calculate updates
    updates = {}
    if payload.trip_name is not None:
        updates["trip_name"] = payload.trip_name.strip() if payload.trip_name else None
    if payload.visited_at is not None:
        try:
            from datetime import datetime
            iso_str = payload.visited_at
            if iso_str.endswith("Z"):
                iso_str = iso_str[:-1] + "+00:00"
            datetime.fromisoformat(iso_str)
            formatted_visited_at = payload.visited_at
            if not formatted_visited_at.endswith("Z") and "+" not in formatted_visited_at:
                formatted_visited_at += "Z"
            updates["visited_at"] = formatted_visited_at
        except Exception:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="visited_at must be a valid ISO datetime string."
            )

    if updates:
        await db.checkins.update_one({"id": id, "user_id": user_id}, {"$set": updates})

    new_trip_name = updates.get("trip_name", old_trip_name)

    # Recalculate trip distances if trip changed or order shifted
    if old_trip_name:
        await recalculate_trip_distances(user_id, old_trip_name)
    if new_trip_name and new_trip_name != old_trip_name:
        await recalculate_trip_distances(user_id, new_trip_name)

    updated = await db.checkins.find_one({"id": id, "user_id": user_id}, {"_id": 0})
    return updated

@router.patch("/assign-trip")
async def assign_trip(
    payload: AssignTripPayload,
    current_user: dict = Depends(get_current_user)
):
    db = get_db()
    user_id = current_user["id"]
    clean_trip_name = payload.trip_name.strip() if payload.trip_name else None

    # Fetch pre-existing trip names to recalculate
    cursor = db.checkins.find({"id": {"$in": payload.checkin_ids}, "user_id": user_id})
    affected = await cursor.to_list(length=None)
    
    old_trips = list(set([c.get("trip_name") for c in affected if c.get("trip_name")]))

    # Batch update
    await db.checkins.update_many(
        {"id": {"$in": payload.checkin_ids}, "user_id": user_id},
        {"$set": {"trip_name": clean_trip_name}}
    )

    # Recalculate distances for newly assigned and all old trips
    if clean_trip_name:
        await recalculate_trip_distances(user_id, clean_trip_name)
    for old_trip in old_trips:
        if old_trip != clean_trip_name:
            await recalculate_trip_distances(user_id, old_trip)

    return {"message": "Stops assigned and distances calculated successfully."}
